server/aesdsocket.c

```diff
@@ -134,71 +219,228 @@ static int echo_file_socket(int fd,int read_len);
+    socket_state.itime.it_interval.tv_sec = 0;
+    socket_state.itime.it_interval.tv_nsec = 0;
+    socket_state.itime.it_value.tv_sec = 0;
+    socket_state.itime.it_value.tv_nsec = 0;
+    timer_settime(socket_state.timer_1, flags, &socket_state.itime,NULL);
+    timer_delete(socket_state.timer_1);
+}
 static void initialize_socket_state()
 {
 
-    socket_state.clean_addr_info = true;
-    socket_state.clean_socket_descriptor = false;
-    socket_state.clean_socket_fd = false;
-    socket_state.clean_append_fd = false;
-    socket_state.free_append_string = false;
-    socket_state.run_as_daemon = false;
-    socket_state.reading_data = false;
+    socket_state.free_address_info = false;
+    socket_state.free_socket_descriptor = false;
     socket_state.signal_caught = false;
+    socket_state.disarm_alarm = false;
     socket_state.host_addr_info = NULL;
-    socket_state.buf = NULL;
+    pthread_mutex_init(&socket_state.mutex, NULL);
+    socket_state.connection_count = 0;
 }
 static void perform_cleanup()
 {
-    if(socket_state.host_addr_info && socket_state.clean_addr_info)
+    if(socket_state.host_addr_info && socket_state.free_address_info)
     {
         freeaddrinfo(socket_state.host_addr_info);
     }
-    if(socket_state.clean_socket_descriptor)
+    if(socket_state.free_address_info)
     {
         close(socket_state.socket_descriptor);
     }
-    if(socket_state.clean_socket_fd)
-    {
-        close(socket_state.socket_file_descriptor);
-    }
-    if(socket_state.clean_append_fd)
-    {
-        close(socket_state.append_file_descriptor);
-    }
-    if(socket_state.free_append_string)
+    if(socket_state.disarm_alarm)
     {
-        free(socket_state.buf);
+        disarm_and_destroy_timer();
     }
+    pthread_mutex_destroy(&socket_state.mutex);
+    //Disarm Alarm
     closelog();
 }
 static void shutdown_function()
 {
     printf("\nCaught Signal. Exiting\n");
     perform_cleanup();
-    if(socket_state.clean_append_fd)
-    {
-        printf("Deleting file\n");
-        unlink("/var/tmp/aesdsocketdata");
-    }
+    printf("Deleting file\n");
+    unlink("/var/tmp/aesdsocketdata");
     exit(1);
 }
 
 static void sighandler()
 {
-    
-    if(!socket_state.reading_data)
-    {
-        shutdown_function();
-    }
-    else
-    {
-        socket_state.signal_caught = true;
-    }
+    socket_state.signal_caught = true;
+}
 
+/*
+*   Functions time() and localtime_r are reentrant and hence are safe to call from signal handler.
+*   strftime() has not been called since it is not reentrant.
+*   Limitations of circular buffer have been documented above.
+*/
+static void alarmhandler()
+{
+    time_t rawtime;
+    struct tm info;
+    char buffer[80];
+    time( &rawtime );
+    localtime_r( &rawtime,&info );
+    sprintf(buffer,"timestamp: %d/%02d/%02d - %02d:%02d:%02d\n",(info.tm_year + 1900),(info.tm_mon + 1),info.tm_mday,info.tm_hour,info.tm_min,info.tm_sec);
+    enqueue_data_into_circ_buf(buffer);
```

Comment:
@GuruprashanthKrishnakumar interesting solution and I like the detail in the description regarding reentrant checks.  I'm not sure you can count on sprintf being reentrant, however.

A simpler solution is probably to use `SIGEV_THREAD` with your timer function

Reply:
Should I have used a thread just for documenting the time or for the time writeback itself? I was concerned that if I used SIGEV_THREAD for writeback, if the thread preempts a server thread which is already holding a lock for the file, it will lead to a deadlock.

Reply:
> I was concerned that if I used SIGEV_THREAD for writeback, if the thread preempts a server thread which is already holding a lock for the file, it will lead to a deadlock.

This is the benefit of the threaded rather than the signal based implementation - threads won't run to completion in a preemption scenario, and can sleep to allow other threads to run.  So if the thread attempts to get the lock for the file and the lock isn't available it will just sleep until it's available.

Reply:
oh I was not aware of that. Thanks!